spec-tests/src/test/java/org/candlepin/spec/consumers/ConsumerCheckinSpecTest.java

```diff
@@ -0,0 +1,213 @@
+        ApiClient consumerClient = ApiClients.ssl(consumer);
+        OffsetDateTime initialCheckin = consumer.getLastCheckin();
+
+        hypervisorCheckin(owner, consumerClient, StringUtil.random("name-"), StringUtil.random("id-"),
+            List.of("guest-1", "guest-2", "guest-3"), Map.of("test_fact", "fact_value"),
+            StringUtil.random("reporter-"), true);
+
+        consumer = consumerClient.consumers().getConsumer(consumer.getUuid());
+        assertThat(consumer)
+            .isNotNull()
+            .doesNotReturn(initialCheckin, ConsumerDTO::getLastCheckin);
+    }
+
+    @Test
+    public void shouldNotUpdateLastCheckinWhenFetchingReleases() throws Exception {
+        ApiClient adminClient = ApiClients.admin();
+        OwnerDTO owner = adminClient.owners().createOwner(Owners.random());
+        ConsumerDTO consumer = adminClient.consumers().createConsumer(Consumers.random(owner));
+        ApiClient consumerClient = ApiClients.ssl(consumer);
+        OffsetDateTime initialCheckin = consumer.getLastCheckin();
+
+        consumerClient.consumers().getRelease(consumer.getUuid());
+
+        consumer = consumerClient.consumers().getConsumer(consumer.getUuid());
+        assertThat(consumer)
+            .isNotNull()
+            .returns(initialCheckin, ConsumerDTO::getLastCheckin);
+    }
+
+    private HypervisorUpdateResultDTO hypervisorCheckin(OwnerDTO owner, ApiClient consumerClient,
+        String hypervisorName, String hypervisorId, List<String> guestIds, Map<String, String> facts,
+        String reporterId, boolean createMissing) throws ApiException, IOException {
+        JsonNode hostGuestMapping = getAsyncHostGuestMapping(hypervisorName, hypervisorId, guestIds, facts);
+        AsyncJobStatusDTO job = consumerClient.hypervisors().hypervisorUpdateAsync(
+            owner.getKey(), createMissing, reporterId, hostGuestMapping.toString());
+        HypervisorUpdateResultDTO resultData = null;
+        if (job != null) {
+            AsyncJobStatusDTO status = consumerClient.jobs().waitForJob(job.getId());
+            assertThatJob(status).isFinished();
+            resultData = getResultData(status);
+        }
+
+        return resultData;
+    }
+
+    private JsonNode getAsyncHostGuestMapping(String name, String id, List<String> expectedGuestIds,
+        Map<String, String> facts) {
+        ObjectMapper om = ApiClient.MAPPER;
+        ArrayNode guestIds = om.createArrayNode();
+        expectedGuestIds.stream()
+            .map(gid -> om.createObjectNode().put("guestId", gid))
+            .forEach(guestIds::add);
+
+        ObjectNode root = om.createObjectNode();
+        ArrayNode hypervisors = root.putArray("hypervisors");
+        ObjectNode hypervisor = om.createObjectNode()
+            .put("name", name)
+            .set("hypervisorId", om.createObjectNode()
+            .put("hypervisorId", id));
+
+        hypervisor.set("guestIds", guestIds);
```

Comment:
Looks like this got partially reverted...?

The object->array for the root can also be chained. The "special" `putArray` and `putObject` function by adding the property and the switching to the new container, so the final wrapper of wrappers can be created in-line:

```
return mapper.createObjectNode()
  .putArray("hypervisors")
  .add(hypervisor);
```

When it comes to building objects like this, it's better for maintenance to ensure the order is as linear as possible; either building it in full reverse order, or top-down. The more it jumps around, the harder it is to visualize the structure as it's being constructed.

Reply:
That makes sense! I did have to make a couple changes. I couldn't chain the `.set()` calls when building the hypervisor ObjectNode as the `.set()` method returns a JsonNode and not an ObjectNode which is strange, and a quick search online shows others aren't too happy about this as well. 

Also, I used `.putPOJO("hypervisors", mapper.createArrayNode().add(hypervisor))` because it returns the outer wrapper object. The example you gave returns the ArrayNode instead of the Object Node and you get something like: 
`[{"name":"name-35008A87","hypervisorId":{"hypervisorId":"id-AF9BA895"},"guestIds":[{"guestId":"guest-1"},{"guestId":"guest-2"},{"guestId":"guest-3"}],"facts":{"test_fact":"fact_value"}}]`

Reply:
> That makes sense! I did have to make a couple changes. I couldn't chain the `.set()` calls when building the hypervisor ObjectNode as the `.set()` method returns a JsonNode and not an ObjectNode which is strange, and a quick search online shows others aren't too happy about this as well.

Hrmm... that's pretty odd. Especially since ObjectNode is what defines `.set` to begin with.

I'm beginning to wonder if building it with Jackson objects is the way to go here, then. We could just as easily build it up with a bunch of maps and lists, and then throw that at Jackson:

```
  List<Map<String, String>> guestIds = expectedGuestIds.stream()
      .map(gid -> Map.of("guestId", gid))
      .collect(Collectors.toList());

  Map<String, Object> hypervisor = Map.of(
      "name", name,
      "hypervisorId", Map.of("hypervisorId", id),
      "guestIds", guestIds,
      "facts", facts);

  Object node = Map.of("hypervisors", List.of(hypervisor));

  return mapper.valueToTree(node);
```


> The example you gave returns the ArrayNode instead of the Object Node and you get something like:

Oops, good catch.

In this case, however, since the output specifies a JsonNode, the detrimental behavior of `.set` isn't an issue as we no longer need the ObjectNode specification.

Reply:
+1 to building with maps and lists! I like that so much more and is easy to read as well. Updated